spec/firebase_app_distribution_action_spec.rb

```diff
@@ -1,141 +1,214 @@
-    end
-
-    it 'removes trailing newlines from firebase_cli_path' do
-      expect(Fastlane::Actions).to receive(:sh_control_output).with("/tmp/fake-firebase-cli appdistribution:distribute /tmp/FakeApp.ipa --app abc:123 --testers-file /tmp/testers.txt --release-notes-file /tmp/release_notes.txt --token fake-token --debug", { print_command: false, print_command_output: true })
-      params = {
-          app: "abc:123",
-          ipa_path: "/tmp/FakeApp.ipa",
-          firebase_cli_path: "/tmp/fake-firebase-cli\n",
-          testers_file: "/tmp/testers.txt",
-          release_notes_file: "/tmp/release_notes.txt",
-          firebase_cli_token: "fake-token",
-          debug: true
-      }
-      Fastlane::Actions::FirebaseAppDistributionAction.run(params)
-    end
-  end
-
-  describe "flag helpers" do
-    let(:action) { Fastlane::Actions::FirebaseAppDistributionAction }
-
-    describe "flag_value_if_supplied" do
-      it "returns flag and param value when it exists" do
-        params = {
-            firebase_cli_token: 'fake-token'
-        }
-        expect(action.flag_value_if_supplied('--token', :firebase_cli_token, params)).to eq("--token fake-token")
-      end
-
-      it "returns nil if param does not exist" do
-        params = {}
-        expect(action.flag_value_if_supplied('--token', :firebåse_cli_token, params)).to be_nil
+  context '#get_upload_token' do
+    context 'when testing with valid parameters' do
+      it 'should make a GET call to the app endpoint and return the upload token' do
+        expect(temp_connection).to receive(:get).with("/v1alpha/apps/app_id").and_return(fake_get_app_response_valid)
+        upload_token = Fastlane::Actions::FirebaseAppDistributionAction.get_upload_token("app_id", "binary_path")
+        binary_hash = Digest::SHA256.hexdigest(File.open(fake_binary).read)
+        expect(upload_token).to eq(CGI.escape("projects/1/apps/1/releases/-/binaries/#{binary_hash}"))
       end
     end
 
-    describe "flag_if_supplied" do
-      it "returns flag and param value when it exists" do
-        params = {
-            debug: true
-        }
-        expect(action.flag_if_supplied('--debug', :debug, params)).to eq("--debug")
+    context 'when testing with invalid parameters' do
+      it 'should crash if the app has no contact email' do
+        expect(temp_connection).to receive(:get).with("/v1alpha/apps/app_id").and_return(fake_get_app_response_no_email)
+        expect { Fastlane::Actions::FirebaseAppDistributionAction.get_upload_token("app_id", "binary_path") }.to raise_error(ErrorMessage::GET_APP_NO_CONTACT_EMAIL_ERROR)
       end
 
-      it "returns nil if param does not exist" do
-        params = {}
-        expect(action.flag_if_supplied('--debug', :debug, params)).to be_nil
+      it 'should crash if given an invalid app_id' do
+        expect(temp_connection).to receive(:get).with("/v1alpha/apps/invalid_app_id").and_raise(Faraday::ResourceNotFound.new("404"))
```

Comment:
Try to format these long lines so they are easier to read. For example `.with(...` and `.and_raise(...` can each drop to the next line.